### ISSUE TITLE
Problems with (IPv6?) tunnel interface statistics?

### ISSUE THREAD

**Issue by hvisage** [vergoh/vnstat#96]
Hi there,

like to start breaking up and see my IPv6 tunnel's traffic usage, I thought I could use vnstat, so last night "enable/created" the needed databases, but seems it doesn't:

```
root@hvs:~/vnstat-1.18# vnstat -q -i he-ipv6
 he-ipv6: Not enough data available yet.
root@hvs:~/vnstat-1.18# vnstat -q -i he-ipv6\@none
Error: Unable to read database "/var/lib/vnstat/he-ipv6@none": No such file or directory
root@hvs:~/vnstat-1.18# vnstat -q -i he-ipv6 -d
 he-ipv6: Not enough data available yet.
root@hvs:~/vnstat-1.18# vnstat -q -i he-ipv6 -h
 he-ipv6: Not enough data available yet.
root@hvs:~/vnstat-1.18# vnstat -q -i he-ipv6 -w
 he-ipv6: Not enough data available yet.
root@hvs:~/vnstat-1.18# vnstat -q -i he-ipv6 -w
 he-ipv6: Not enough data available yet.
root@hvs:~/vnstat-1.18#
```

```
auto he-ipv6
iface he-ipv6 inet6 v4tunnel
        address 2001:4xx:xxx:3d::2
        netmask 64
        endpoint 216.66.87.134
        ttl 255
        gateway 2001:4xx:xxx:3d::1
```

```
root@hvs:~/vnstat-1.18# ip link show dev he-ipv6
12: he-ipv6@NONE: <POINTOPOINT,NOARP,UP,LOWER_UP> mtu 1472 qdisc noqueue state UNKNOWN mode DEFAULT group default
    link/sit 0.0.0.0 peer 216.xxx.xxx.xxx
root@hvs:~/vnstat-1.18# ifconfig he-ipv6
he-ipv6   Link encap:IPv6-in-IPv4
          inet6 addr: fe80::ac12:f001/64 Scope:Link
          inet6 addr: fe80::a5ff:ba92/64 Scope:Link
          inet6 addr: 2001:4xx:xxx:3d::2/64 Scope:Global
          inet6 addr: fe80::a0a:a02/64 Scope:Link
          inet6 addr: fe80::a01:101/64 Scope:Link
          UP POINTOPOINT RUNNING NOARP  MTU:1472  Metric:1
          RX packets:749839 errors:0 dropped:0 overruns:0 frame:0
          TX packets:659729 errors:0 dropped:0 overruns:0 carrier:0
          collisions:0 txqueuelen:0
          RX bytes:445951666 (425.2 MiB)  TX bytes:75758521 (72.2 MiB)


root@hvs:~/vnstat-1.18# vnstat -i he-ipv6 --dumpdb
version;3
active;0
interface;he-ipv6
nick;he-ipv6
created;1523392052
updated;1523393424
totalrx;0
totaltx;0
currx;0
curtx;0
totalrxk;0
totaltxk;0
btime;4294967295
d;0;1523392052;0;0;0;0;1
d;1;0;0;0;0;0;0
d;2;0;0;0;0;0;0
d;3;0;0;0;0;0;0
```

**Comment by vergoh** [comment]
That's a little bit strange that vnStat doesn't see any traffic but the `ifconfig` output clearly shows that the necessary data should be available somewhere. Few questions that could help understanding what's going on:

1. Do those rx and tx bytes shown in the `ifconfig` output match exactly those of the interface itself that is being used by the tunnel on the ipv4 side or are the numbers clearly different?
2. Is he-ipv6 visible in `/proc/net/dev` with non-zero numbers?
3. Does `/sys/class/net/em1/statistics/rx_bytes` exist with a non-zero content?
4. Which distribution is being used?
5. Is grsecurity or some other kernel security enhancement enabled?

**Comment by hvisage** [comment]
1) I can't say. The traffic would be something like: IPv6 -> he-ipv6 -> PPPx -> eth2 -> ADSL modem
I'm also now notice that some of the other tunnel/VPN interfaces (tma/cloudnet, both tincd based) doesn't have "enough information" either

2 & 3) yes. even the tincd based interfaces, see atttachments
[proc-net-dev.txt](https://github.com/vergoh/vnstat/files/1900416/proc-net-dev.txt)
[sys-class-net.txt](https://github.com/vergoh/vnstat/files/1900421/sys-class-net.txt)
[sys-class-net-statis.txt](https://github.com/vergoh/vnstat/files/1900417/sys-class-net-statis.txt)

4) Devuan. `Linux hvs 3.18.9-031809-generic #201503080036 SMP Sun Mar 8 00:46:02 UTC 2015 i686 GNU/Linux`

5) not what I'm aware of

**Comment by vergoh** [comment]
Thanks for the output files. I tested the `/proc/net/dev` output file with a modified version of both 1.18 and 2.0 beta that were updated to read the file instead of the actual `/proc/net/dev` path and neither had problems detecting he-ipv6 when `vnstat --iflist` was used or when `vnstat -tr -i he-ipv6` was executed. Obviously, the later didn't report any traffic but the debug output shows that the problem isn't in the parsing of the interface data when it can be accessed.

Could you provide the outputs of

- `vnstat -D --iflist`
- `vnstat -D -tr -i he-ipv6`

Also, I noticed that although you executed the command of the first message from ~/vnstat-1.18, the vnstat binary was from the PATH instead of the current directory. Does `vnstat --version` show that the version that is being used is indeed 1.18?

**Comment by hvisage** [comment]
Let's start from the bottom: I compiled 1.18 as the version in Devuan was a 1.12 something.
```
root@hvs:~# vnstat --version
vnStat 1.18 by Teemu Toivola <tst at iki dot fi>
root@hvs:~# which vnstat
/usr/bin/vnstat
root@hvs:~#
```

- vnstat -D --iflist
[iflist.txt](https://github.com/vergoh/vnstat/files/1905229/iflist.txt)

- vnstat -D -tr -i he-ipv6
 [tr-he-ipv6.txt](https://github.com/vergoh/vnstat/files/1905230/tr-he-ipv6.txt)

seems the -D -tr was able to see traffic as I was youtubing...

**Comment by vergoh** [comment]
Ok, the `-D -tr -i he-ipv6` output does indeed show the traffic so that rules out the possibility of this being some sort of parsing issue when getting the traffic data.

Did you have the 1.12 package from Devuan previously installed before replacing it with 1.18? If yes, was the previous package uninstalled first before `make install` with 1.18?

In the first comment, you mention that you "enable/created" the needed databases. Usually the daemon automatically creates all databases based on found interfaces unless some databases were already there from a previous install. How has the daemon been configured to be started and is it still running?

**Comment by vergoh** [comment]
Sorry, I got somewhat mislead by the existence of the v4tunnel device. This is the first report of such usage I've seen and based on previous experiences, tunnel devices may require some modifications or lack the statistics needed by vnStat. That's why I didn't first think of asking the normal battery of questions related to the used OS, configuration, daemon, file permissions etc. (probably should make a issue report template of this).

Anyway, good that you figured the permission issue. There should be some sort of complaint in the syslog regarding the daemon not being able to write to the directory. The problem itself is caused by the init script which starts the daemon directly as the vnstat user instead of relaying this information to the root started daemon which would the fix any permissions issues before falling back to the requested user and group. Since the init script wants to be in control, it should also have the ability of fixing file permissions.

This file permission issue is, in a way, fixed in version 2.0 (currently beta) which replaces the interface specific database files with one sqlite database. The daemon will not start unless it has write access to the database and any user modifications (with enough permissions) will not break the permissions either since the permissions of an existing file doesn't change on modification.
